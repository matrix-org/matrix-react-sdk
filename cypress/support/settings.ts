/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/// <reference types="cypress" />

import Chainable = Cypress.Chainable;
import Loggable = Cypress.Loggable;
import Timeoutable = Cypress.Timeoutable;
import Withinable = Cypress.Withinable;
import Shadow = Cypress.Shadow;
import type { SettingLevel } from "../../src/settings/SettingLevel";
import type SettingsStore from "../../src/settings/SettingsStore";

export enum Filter {
    People = "people",
    PublicRooms = "public_rooms",
}

declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace Cypress {
        interface Chainable {
            /**
             * Returns the SettingsStore
             */
            getSettingsStore(): Chainable<typeof SettingsStore | undefined>; // XXX: Importing SettingsStore causes a bunch of type lint errors
            /**
             * Open the top left user menu, returning a handle to the resulting context menu.
             */
            openUserMenu(): Chainable<JQuery<HTMLElement>>;

            /**
             * Open user settings (via user menu), returning a handle to the resulting dialog.
             * @param tab the name of the tab to switch to after opening, optional.
             */
            openUserSettings(tab?: string): Chainable<JQuery<HTMLElement>>;

            /**
             * Open room creation dialog.
             */
            openCreateRoomDialog(): Chainable<JQuery<HTMLElement>>;

            /**
             * Open room settings (via room header menu), returning a handle to the resulting dialog.
             * @param tab the name of the tab to switch to after opening, optional.
             */
            openRoomSettings(tab?: string): Chainable<JQuery<HTMLElement>>;

            /**
             * Switch settings tab to the one by the given name, ideally call this in the context of the dialog.
             * @param tab the name of the tab to switch to.
             */
            switchTab(tab: string): Chainable<JQuery<HTMLElement>>;

            /**
             * Close dialog, ideally call this in the context of the dialog.
             */
            closeDialog(): Chainable<JQuery<HTMLElement>>;

            /**
             * Join the given beta, the `Labs` tab must already be opened,
             * ideally call this in the context of the dialog.
             * @param name the name of the beta to join.
             */
            joinBeta(name: string): Chainable<JQuery<HTMLElement>>;

            /**
             * Leave the given beta, the `Labs` tab must already be opened,
             * ideally call this in the context of the dialog.
             * @param name the name of the beta to leave.
             */
            leaveBeta(name: string): Chainable<JQuery<HTMLElement>>;

            /**
             * Sets the value for a setting. The room ID is optional if the
             * setting is not being set for a particular room, otherwise it
             * should be supplied. The value may be null to indicate that the
             * level should no longer have an override.
             * @param {string} settingName The name of the setting to change.
             * @param {String} roomId The room ID to change the value in, may be
             * null.
             * @param {SettingLevel} level The level to change the value at.
             * @param {*} value The new value of the setting, may be null.
             * @return {Promise} Resolves when the setting has been changed.
             */
            setSettingValue(settingName: string, roomId: string, level: SettingLevel, value: any): Chainable<void>;

            /**
             * Gets the value of a setting. The room ID is optional if the
             * setting is not to be applied to any particular room, otherwise it
             * should be supplied.
             * @param {string} settingName The name of the setting to read the
             * value of.
             * @param {String} roomId The room ID to read the setting value in,
             * may be null.
             * @param {boolean} excludeDefault True to disable using the default
             * value.
             * @return {*} The value, or null if not found
             */
            getSettingValue<T>(settingName: string, roomId?: string, excludeDefault?: boolean): Chainable<T>;

            /**
             * Opens the spotlight dialog
             */
            openSpotlightDialog(
                options?: Partial<Loggable & Timeoutable & Withinable & Shadow>,
            ): Chainable<JQuery<HTMLElement>>;
            spotlightDialog(
                options?: Partial<Loggable & Timeoutable & Withinable & Shadow>,
            ): Chainable<JQuery<HTMLElement>>;
            spotlightFilter(
                filter: Filter | null,
                options?: Partial<Loggable & Timeoutable & Withinable & Shadow>,
            ): Chainable<JQuery<HTMLElement>>;
            spotlightSearch(
                options?: Partial<Loggable & Timeoutable & Withinable & Shadow>,
            ): Chainable<JQuery<HTMLElement>>;
            spotlightResults(
                options?: Partial<Loggable & Timeoutable & Withinable & Shadow>,
            ): Chainable<JQuery<HTMLElement>>;
        }
    }
}

Cypress.Commands.add("getSettingsStore", (): Chainable<typeof SettingsStore> => {
    return cy.window({ log: false }).then((win) => win.mxSettingsStore);
});

Cypress.Commands.add(
    "setSettingValue",
    (name: string, roomId: string, level: SettingLevel, value: any): Chainable<void> => {
        return cy.getSettingsStore().then((store: typeof SettingsStore) => {
            return cy.wrap(store.setValue(name, roomId, level, value));
        });
    },
);

// eslint-disable-next-line max-len
Cypress.Commands.add(
    "getSettingValue",
    <T = any>(name: string, roomId?: string, excludeDefault?: boolean): Chainable<T> => {
        return cy.getSettingsStore().then((store: typeof SettingsStore) => {
            return store.getValue(name, roomId, excludeDefault);
        });
    },
);

Cypress.Commands.add("openUserMenu", (): Chainable<JQuery<HTMLElement>> => {
    cy.findByRole("button", { name: "User menu" }).click();
    return cy.get(".mx_ContextualMenu");
});

Cypress.Commands.add("openUserSettings", (tab?: string): Chainable<JQuery<HTMLElement>> => {
    cy.openUserMenu().within(() => {
        cy.findByRole("menuitem", { name: "All settings" }).click();
    });
    return cy.get(".mx_UserSettingsDialog").within(() => {
        if (tab) {
            cy.switchTab(tab);
        }
    });
});

Cypress.Commands.add("openCreateRoomDialog", (): Chainable<JQuery<HTMLElement>> => {
    cy.findByRole("button", { name: "Add room" }).click();
    cy.findByRole("menuitem", { name: "New room" }).click();
    return cy.get(".mx_CreateRoomDialog");
});

Cypress.Commands.add("openRoomSettings", (tab?: string): Chainable<JQuery<HTMLElement>> => {
    cy.findByRole("button", { name: "Room options" }).click();
    cy.get(".mx_RoomTile_contextMenu").within(() => {
        cy.findByRole("menuitem", { name: "Settings" }).click();
    });
    return cy.get(".mx_RoomSettingsDialog").within(() => {
        if (tab) {
            cy.switchTab(tab);
        }
    });
});

Cypress.Commands.add("switchTab", (tab: string): Chainable<JQuery<HTMLElement>> => {
    return cy.get(".mx_TabbedView_tabLabels").within(() => {
        cy.contains(".mx_TabbedView_tabLabel", tab).click();
    });
});

Cypress.Commands.add("closeDialog", (): Chainable<JQuery<HTMLElement>> => {
    return cy.findByRole("button", { name: "Close dialog" }).click();
});

Cypress.Commands.add("joinBeta", (name: string): Chainable<JQuery<HTMLElement>> => {
    return cy
        .contains(".mx_BetaCard_title", name)
        .closest(".mx_BetaCard")
        .within(() => {
            return cy.get(".mx_BetaCard_buttons").findByRole("button", { name: "Join the beta" }).click();
        });
});

Cypress.Commands.add("leaveBeta", (name: string): Chainable<JQuery<HTMLElement>> => {
    return cy
        .contains(".mx_BetaCard_title", name)
        .closest(".mx_BetaCard")
        .within(() => {
            return cy.get(".mx_BetaCard_buttons").findByRole("button", { name: "Leave the beta" }).click();
        });
});

Cypress.Commands.add(
    "openSpotlightDialog",
    (options?: Partial<Loggable & Timeoutable & Withinable & Shadow>): Chainable<JQuery<HTMLElement>> => {
        cy.get(".mx_RoomSearch_spotlightTrigger", options).click({ force: true });
        return cy.spotlightDialog(options);
    },
);

Cypress.Commands.add(
    "spotlightDialog",
    (options?: Partial<Loggable & Timeoutable & Withinable & Shadow>): Chainable<JQuery<HTMLElement>> => {
        return cy.get('[role=dialog][aria-label="Search Dialog"]', options);
    },
);

Cypress.Commands.add(
    "spotlightFilter",
    (
        filter: Filter | null,
        options?: Partial<Loggable & Timeoutable & Withinable & Shadow>,
    ): Chainable<JQuery<HTMLElement>> => {
        let selector: string;
        switch (filter) {
            case Filter.People:
                selector = "#mx_SpotlightDialog_button_startChat";
                break;
            case Filter.PublicRooms:
                selector = "#mx_SpotlightDialog_button_explorePublicRooms";
                break;
            default:
                selector = ".mx_SpotlightDialog_filter";
                break;
        }
        return cy.get(selector, options).click();
    },
);

Cypress.Commands.add(
    "spotlightSearch",
    (options?: Partial<Loggable & Timeoutable & Withinable & Shadow>): Chainable<JQuery<HTMLElement>> => {
        return cy.get(".mx_SpotlightDialog_searchBox", options).findByRole("textbox", { name: "Search" });
    },
);

Cypress.Commands.add(
    "spotlightResults",
    (options?: Partial<Loggable & Timeoutable & Withinable & Shadow>): Chainable<JQuery<HTMLElement>> => {
        return cy.get(".mx_SpotlightDialog_section.mx_SpotlightDialog_results .mx_SpotlightDialog_option", options);
    },
);

// Needed to make this file a module
export {};
