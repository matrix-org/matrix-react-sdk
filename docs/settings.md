# Settings Reference

This document serves as developer documentation for using "Granular Settings". Granular Settings allow users to specify different values for a setting at particular levels of interest. For example, a user may say that in a particular room they want URL previews off, but in all other rooms they want them enabled. The `SettingsStore` helps mask the complexity of dealing with the different levels and exposes easy to use getters and setters.


## Levels

Granular Settings rely on a series of known levels in order to use the correct value for the scenario. These levels, in order of prioirty, are:
* `device` - The current user's device
* `room-device` - The current user's device, but only when in a specific room
* `room-account` - The current user's account, but only when in a specific room
* `account` - The current user's account
* `room` - A specific room (setting for all members of the room)
* `config` - Values are defined by `config.json`
* `default` - The hardcoded default for the settings

Individual settings may control which levels are appropriate for them as part of the defaults. This is often to ensure that room administrators cannot force account-only settings upon participants.


## Settings

Settings are the different options a user may set or experience in the application. These are pre-defined in `src/settings/Settings.js` under the `SETTINGS` constant and have the following minimum requirements:
```
// The ID is used to reference the setting throughout the application. This must be unique.
"theSettingId": {
  // The levels this setting supports is required. In `src/settings/Settings.js` there are various pre-set arrays
  // for this option - they should be used where possible to avoid copy/pasting arrays across settings.
  supportedLevels: [...],

  // The default for this setting serves two purposes: It provides a value if the setting is not defined at other
  // levels, and it serves to demonstrate the expected type to other developers. The value isn't enforced, but it
  // should be respected throughout the code. The default may be any data type.
  default: false,

  // The display name has two notations: string and object. The object notation allows for different translatable
  // strings to be used for different levels, while the string notation represents the string for all levels.

  displayName: _td("Change something"), // effectively `displayName: { "default": _td("Change something") }`
  displayName: {
    "room": _td("Change something for participants of this room"),

    // Note: the default will be used if the level requested (such as `device`) does not have a string defined here.
    "default": _td("Change something"),
  }
}
```

### Getting values for a setting

After importing `SettingsStore`, simply make a call to `SettingsStore.getValue`. The `roomId` parameter should always be supplied where possible, even if the setting does not have a per-room level value. This is to ensure that the value returned is best represented in the room, particularly if the setting ever gets a per-room level in the future.

In settings pages it is often desired to have the value at a particular level instead of getting the calculated value. Call `SettingsStore.getValueAt` to get the value of a setting at a particular level, and optionally make it explicitly at that level. By default `getValueAt` will traverse the tree starting at the provided level; making it explicit means it will not go beyond the provided level. When using `getValueAt`, please be sure to use `SettingLevel` to represent the target level.

### Setting values for a setting

Values are defined at particular levels and should be done in a safe manner. There are two checks to perform to ensure a clean save: is the level supported and can the user actually set the value. In most cases, neither should be an issue although there are circumstances where this changes. An example of a safe call is:
```javascript
const isSupported = SettingsStore.isLevelSupported(SettingLevel.ROOM);
if (isSupported) {
  const canSetValue = SettingsStore.canSetValue("mySetting", "!curbf:matrix.org", SettingLevel.ROOM);
  if (canSetValue) {
    SettingsStore.setValue("mySetting", "!curbf:matrix.org", SettingLevel.ROOM, newValue);
  }
}
```

These checks may also be performed in different areas of the application to avoid the verbose example above. For instance, the component which allows changing the setting may be hidden conditionally on the above conditions.

##### `SettingsFlag` component

Where possible, the `SettingsFlag` component should be used to set simple "flip-a-bit" (true/false) settings. The `SettingsFlag` also supports simple radio button options, such as the theme the user would like to use.
```html
<SettingsFlag name="theSettingId"
              level={SettingsLevel.ROOM}
              roomId="!curbf:matrix.org"
              label={_td("Your label here")} // optional, if falsey then the `SettingsStore` will be used
              onChange={function(newValue) { }} // optional, called after saving
              isExplicit={false} // this is passed along to `SettingsStore.getValueAt`, defaulting to false
              manualSave={false} // if true, saving is delayed. You will need to call .save() on this component
              
              // Options for radio buttons
              group="your-radio-group" // this enables radio button support
              value="yourValueHere" // the value for this particular option
/>
```

### Getting the display name for a setting

Simply call `SettingsStore.getDisplayName`. The appropriate display name will be returned and automatically translated for you. If a display name cannot be found, it will return `null`.


## Features

Occasionally some parts of the application may be undergoing testing and are not quite production ready. These are commonly known to be behind a "labs flag". Features behind lab flags must go through the granular settings system, and look and act very much normal settings. The exception is that they must supply `isFeature: true` as part of the setting definition and should go through the helper functions on `SettingsStore`.

Although features have levels and a default value, the calculation of those options is blocked by the feature's state. A feature's state is determined from the `SdkConfig` and is a little complex. If `enableLabs` (a legacy flag) is `true` then the feature's state is `labs`, if it is `false`, the state is `disable`. If `enableLabs` is not set then the state is determined from the `features` config, such as in the following:
```json
"features": {
    "feature_lazyloading": "labs"
}
```
In this example, `feature_lazyloading` is in the `labs` state. It may also be in the `enable` or `disable` state with a similar approach. If the state is invalid, the feature is in the `disable` state. A feature's levels are only calculated if it is in the `labs` state, therefore the default only applies in that scenario. If the state is `enable`, the feature is always-on.

Once a feature flag has served its purpose, it is generally recommended to remove it and the associated feature flag checks. This would enable the feature implicitly as it is part of the application now.

### Determining if a feature is enabled

A simple call to `SettingsStore.isFeatureEnabled` will tell you if the feature is enabled. This will perform all the required calculations to determine if the feature is enabled based upon the configuration and user selection.

### Enabling a feature

Features can only be enabled if the feature is in the `labs` state, otherwise this is a no-op. To find the current set of features in the `labs` state, call `SettingsStore.getLabsFeatures`. To set the value, call `SettingsStore.setFeatureEnabled`. 


## Setting controllers

Settings may have environmental factors that affect their value or need additional code to be called when they are modified. A setting controller is able to override the calculated value for a setting and react to changes in that setting. Controllers are not a replacement for the level handlers and should only be used to ensure the environment is kept up to date with the setting where it is otherwise not possible. An example of this is the notification settings: they can only be considered enabled if the platform supports notifications, and enabling notifications requires additional steps to actually enable notifications.

For more information, see `src/settings/controllers/SettingController.js`.


## Local echo

`SettingsStore` will perform local echo on all settings to ensure that immediately getting values does not cause a split-brain scenario. As mentioned in the "Setting values for a setting" section, the appropriate checks should be done to ensure that the user is allowed to set the value. The local echo system assumes that the user has permission and that the request will go through successfully. The local echo only takes effect until the request to save a setting has completed (either successfully or otherwise). 

```javascript
SettingsStore.setValue(...).then(() => {
  // The value has actually been stored at this point.
});
SettingsStore.getValue(...); // this will return the value set in `setValue` above.
```



# Maintainers Reference

The granular settings system has a few complex parts to power it. This section is to document how the `SettingsStore` is supposed to work.

### General information

The `SettingsStore` uses the hardcoded `LEVEL_ORDER` constant to ensure that it is using the correct override procedure. The array is checked from left to right, simulating the behaviour of overriding values from the higher levels. Each level should be defined in this array, including `default`.

Handlers (`src/settings/handlers/SettingsHandler.js`) represent a single level and are responsible for getting and setting values at that level. Handlers also provide additional information to the `SettingsStore` such as if the level is supported or if the current user may set values at the level. The `SettingsStore` will use the handler to enforce checks and manipulate settings. Handlers are also responsible for dealing with migration patterns or legacy settings for their level (for example, a setting being renamed or using a different key from other settings in the underlying store). Handlers are provided to the `SettingsStore` via the `LEVEL_HANDLERS` constant. `SettingsStore` will optimize lookups by only considering handlers that are supported on the platform.

Local echo is achieved through `src/settings/handlers/LocalEchoWrapper.js` which acts as a wrapper around a given handler. This is automatically applied to all defined `LEVEL_HANDLERS` and proxies the calls to the wrapped handler where possible. The echo is achieved by a simple object cache stored within the class itself. The cache is invalidated immediately upon the proxied save call succeeding or failing.

Controllers are notified of changes by the `SettingsStore`, and are given the opportunity to override values after the `SettingsStore` has deemed the value calculated. Controllers are invoked as the last possible step in the code.

### Features

Features automatically get considered as `disabled` if they are not listed in the `SdkConfig` or `enable_labs` is false/not set. Features are always checked against the configuration before going through the level order as they have the option of being forced-on or forced-off for the application. This is done by the `features` section and looks something like this:
 
```
"features": {
  "feature_groups": "enable",
  "feature_pinning": "disable", // the default
  "feature_presence": "labs"
}
```

If `enableLabs` is true in the configuration, the default for features becomes `"labs"`.
