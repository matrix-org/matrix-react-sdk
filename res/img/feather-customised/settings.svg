
<svg width="20px" height="20px" viewBox="0 0 20 20" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <g id="Page-1" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round">
        <g id="LifeBuoy" transform="translate(-1290.000000, -89.000000)" stroke="#B8BEC9" stroke-width="1">
            <g id="settings-copy" transform="translate(1291.000000, 90.000000)">
                <circle id="Oval" cx="9" cy="9" r="2.45454545"></circle>
                <path d="M15.0545455,11.4545455 C14.8317165,11.9594373 14.938637,12.5491199 15.3245455,12.9436364 L15.3736364,12.9927273 C15.6809079,13.2996571 15.85356,13.7161485 15.85356,14.1504545 C15.85356,14.5847606 15.6809079,15.0012519 15.3736364,15.3081818 C15.0667065,15.6154533 14.6502151,15.7881054 14.2159091,15.7881054 C13.781603,15.7881054 13.3651117,15.6154533 13.0581818,15.3081818 L13.0090909,15.2590909 C12.6145745,14.8731825 12.0248919,14.766262 11.52,14.9890909 C11.0254331,15.2010559 10.7039642,15.686474 10.7018182,16.2245455 L10.7018182,16.3636364 C10.7018182,17.267375 9.96919323,18 9.06545455,18 C8.16171586,18 7.42909091,17.267375 7.42909091,16.3636364 L7.42909091,16.29 C7.41612813,15.7358216 7.06571327,15.2458897 6.54545455,15.0545455 C6.04056267,14.8317165 5.45088006,14.938637 5.05636364,15.3245455 L5.00727273,15.3736364 C4.70034285,15.6809079 4.2838515,15.85356 3.84954545,15.85356 C3.41523941,15.85356 2.99874806,15.6809079 2.69181818,15.3736364 C2.38454666,15.0667065 2.21189456,14.6502151 2.21189456,14.2159091 C2.21189456,13.781603 2.38454666,13.3651117 2.69181818,13.0581818 L2.74090909,13.0090909 C3.12681754,12.6145745 3.23373801,12.0248919 3.01090909,11.52 C2.79894413,11.0254331 2.31352603,10.7039642 1.77545455,10.7018182 L1.63636364,10.7018182 C0.732624955,10.7018182 1.81672859e-16,9.96919323 0,9.06545455 C-9.08364293e-17,8.16171586 0.732624955,7.42909091 1.63636364,7.42909091 L1.71,7.42909091 C2.26417842,7.41612813 2.75411031,7.06571327 2.94545455,6.54545455 C3.16828346,6.04056267 3.06136299,5.45088006 2.67545455,5.05636364 L2.62636364,5.00727273 C2.31909211,4.70034285 2.14644002,4.2838515 2.14644002,3.84954545 C2.14644002,3.41523941 2.31909211,2.99874806 2.62636364,2.69181818 C2.93329351,2.38454666 3.34978487,2.21189456 3.78409091,2.21189456 C4.21839695,2.21189456 4.63488831,2.38454666 4.94181818,2.69181818 L4.99090909,2.74090909 C5.38542551,3.12681754 5.97510812,3.23373801 6.48,3.01090909 L6.54545455,3.01090909 C7.04002141,2.79894413 7.36149035,2.31352603 7.36363636,1.77545455 L7.36363636,1.63636364 C7.36363636,0.732624955 8.09626132,1.81672859e-16 9,0 C9.90373868,0 10.6363636,0.732624955 10.6363636,1.63636364 L10.6363636,1.71 C10.6385096,2.24807148 10.9599786,2.73348959 11.4545455,2.94545455 C11.9594373,3.16828346 12.5491199,3.06136299 12.9436364,2.67545455 L12.9927273,2.62636364 C13.2996571,2.31909211 13.7161485,2.14644002 14.1504545,2.14644002 C14.5847606,2.14644002 15.0012519,2.31909211 15.3081818,2.62636364 C15.6154533,2.93329351 15.7881054,3.34978487 15.7881054,3.78409091 C15.7881054,4.21839695 15.6154533,4.63488831 15.3081818,4.94181818 L15.2590909,4.99090909 C14.8731825,5.38542551 14.766262,5.97510812 14.9890909,6.48 L14.9890909,6.54545455 C15.2010559,7.04002141 15.686474,7.36149035 16.2245455,7.36363636 L16.3636364,7.36363636 C17.267375,7.36363636 18,8.09626132 18,9 C18,9.90373868 17.267375,10.6363636 16.3636364,10.6363636 L16.29,10.6363636 C15.7519285,10.6385096 15.2665104,10.9599786 15.0545455,11.4545455 Z" id="Path"></path>
            </g>
        </g>
    </g>
</svg>
