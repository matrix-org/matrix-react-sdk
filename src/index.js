/*
Copyright 2015, 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

var Skinner = require('./Skinner');

var t = require('counterpart');

t.registerTranslations('en', require('../languages/en'));
t.registerTranslations('pt-BR', require('../languages/pt-BR'));

t.setFallbackLocale('en');
t.setLocale(window.navigator.languages[0]);



module.exports.loadSkin = function(skinObject) {
    Skinner.load(skinObject);
};

module.exports.resetSkin = function() {
    Skinner.reset();
};

module.exports.getComponent = function(componentName) {
    return Skinner.getComponent(componentName);
};
