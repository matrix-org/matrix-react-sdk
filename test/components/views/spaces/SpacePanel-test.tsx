/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
// eslint-disable-next-line deprecate/import
import { mount } from 'enzyme';
import { mocked } from 'jest-mock';
import { MatrixClient } from 'matrix-js-sdk/src/matrix';
import { act } from "react-dom/test-utils";

import SpacePanel from '../../../../src/components/views/spaces/SpacePanel';
import { MatrixClientPeg } from '../../../../src/MatrixClientPeg';
import { SpaceKey } from '../../../../src/stores/spaces';
import { findByTestId } from '../../../test-utils';
import { shouldShowComponent } from '../../../../src/customisations/helpers/UIComponents';
import { UIComponent } from '../../../../src/settings/UIFeature';

jest.mock('../../../../src/stores/spaces/SpaceStore', () => {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const EventEmitter = require("events");
    class MockSpaceStore extends EventEmitter {
        invitedSpaces = [];
        enabledMetaSpaces = [];
        spacePanelSpaces = [];
        activeSpace: SpaceKey = '!space1';
    }
    return {
        instance: new MockSpaceStore(),
    };
});

jest.mock('../../../../src/customisations/helpers/UIComponents', () => ({
    shouldShowComponent: jest.fn(),
}));

describe('<SpacePanel />', () => {
    const defaultProps = {};
    const getComponent = (props = {}) =>
        mount(<SpacePanel {...defaultProps} {...props} />);

    const mockClient = {
        getUserId: jest.fn().mockReturnValue('@test:test'),
        isGuest: jest.fn(),
        getAccountData: jest.fn(),
    } as unknown as MatrixClient;

    beforeAll(() => {
        jest.spyOn(MatrixClientPeg, 'get').mockReturnValue(mockClient);
    });

    beforeEach(() => {
        mocked(shouldShowComponent).mockClear().mockReturnValue(true);
    });

    describe('create new space button', () => {
        it('renders create space button when UIComponent.CreateSpaces component should be shown', () => {
            const component = getComponent();
            expect(findByTestId(component, 'create-space-button').length).toBeTruthy();
        });

        it('does not render create space button when UIComponent.CreateSpaces component should not be shown', () => {
            mocked(shouldShowComponent).mockReturnValue(false);
            const component = getComponent();
            expect(shouldShowComponent).toHaveBeenCalledWith(UIComponent.CreateSpaces);
            expect(findByTestId(component, 'create-space-button').length).toBeFalsy();
        });

        it('opens context menu on create space button click', async () => {
            const component = getComponent();

            await act(async () => {
                findByTestId(component, 'create-space-button').at(0).simulate('click');
                component.setProps({});
            });

            expect(component.find('SpaceCreateMenu').length).toBeTruthy();
        });
    });
});
