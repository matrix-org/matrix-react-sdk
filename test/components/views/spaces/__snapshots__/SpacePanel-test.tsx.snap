// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpacePanel /> should allow rearranging via drag and drop 1`] = `
<DocumentFragment>
  <div
    class="mx_SpacePanel collapsed"
  >
    <div
      class="mx_UserMenu"
    >
      <div
        aria-expanded="false"
        aria-haspopup="true"
        aria-label="User menu"
        class="mx_AccessibleButton"
        role="button"
        tabindex="0"
        title="User menu"
      >
        <div
          class="mx_UserMenu_userAvatar"
        >
          <span
            class="mx_BaseAvatar mx_UserMenu_userAvatar_BaseAvatar"
            role="presentation"
          >
            <span
              aria-hidden="true"
              class="mx_BaseAvatar_initial"
              style="font-size: 20.8px; width: 32px; line-height: 32px;"
            >
              T
            </span>
            <img
              alt=""
              aria-hidden="true"
              class="mx_BaseAvatar_image"
              data-testid="avatar-img"
              loading="lazy"
              src="data:image/png;base64,00"
              style="width: 32px; height: 32px;"
            />
          </span>
        </div>
      </div>
      <div
        aria-label="Expand"
        class="mx_AccessibleButton mx_SpacePanel_toggleCollapse"
        role="button"
        tabindex="0"
      />
    </div>
    <ul
      aria-label="Spaces"
      class="mx_AutoHideScrollbar mx_SpaceTreeLevel"
      data-rbd-droppable-context-id="0"
      data-rbd-droppable-id="top-level-spaces"
      role="tree"
      tabindex="-1"
    >
      <li
        aria-selected="false"
        class="mx_SpaceItem mx_SpaceItem_narrow collapsed"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="!room1:server"
        role="treeitem"
      >
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          aria-label="Room 1"
          class="mx_AccessibleButton mx_SpaceButton mx_SpaceButton_narrow"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="!room1:server"
          draggable="false"
          role="button"
          tabindex="0"
        >
          <div
            class="mx_SpaceButton_selectionWrapper"
          >
            <div
              class="mx_SpaceButton_avatarWrapper"
            >
              <span
                class="mx_BaseAvatar"
                role="presentation"
              >
                <span
                  aria-hidden="true"
                  class="mx_BaseAvatar_initial"
                  style="font-size: 20.8px; width: 32px; line-height: 32px;"
                >
                  R
                </span>
                <img
                  alt=""
                  aria-hidden="true"
                  class="mx_BaseAvatar_image"
                  data-testid="avatar-img"
                  loading="lazy"
                  src="data:image/png;base64,00"
                  style="width: 32px; height: 32px;"
                />
              </span>
            </div>
            <div
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Space options"
              class="mx_AccessibleButton mx_SpaceButton_menuButton"
              role="button"
              tabindex="0"
            />
          </div>
        </div>
      </li>
      <li
        aria-selected="false"
        class="mx_SpaceItem mx_SpaceItem_narrow collapsed"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="!room2:server"
        role="treeitem"
      >
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          aria-label="Room 2"
          class="mx_AccessibleButton mx_SpaceButton mx_SpaceButton_narrow"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="!room2:server"
          draggable="false"
          role="button"
          tabindex="-1"
        >
          <div
            class="mx_SpaceButton_selectionWrapper"
          >
            <div
              class="mx_SpaceButton_avatarWrapper"
            >
              <span
                class="mx_BaseAvatar"
                role="presentation"
              >
                <span
                  aria-hidden="true"
                  class="mx_BaseAvatar_initial"
                  style="font-size: 20.8px; width: 32px; line-height: 32px;"
                >
                  R
                </span>
                <img
                  alt=""
                  aria-hidden="true"
                  class="mx_BaseAvatar_image"
                  data-testid="avatar-img"
                  loading="lazy"
                  src="data:image/png;base64,00"
                  style="width: 32px; height: 32px;"
                />
              </span>
            </div>
            <div
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Space options"
              class="mx_AccessibleButton mx_SpaceButton_menuButton"
              role="button"
              tabindex="0"
            />
          </div>
        </div>
      </li>
      <li
        aria-selected="false"
        class="mx_SpaceItem mx_SpaceItem_narrow collapsed"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="!room3:server"
        role="treeitem"
      >
        <div
          aria-describedby="rbd-hidden-text-0-hidden-text-0"
          aria-label="Room 3"
          class="mx_AccessibleButton mx_SpaceButton mx_SpaceButton_narrow"
          data-rbd-drag-handle-context-id="0"
          data-rbd-drag-handle-draggable-id="!room3:server"
          draggable="false"
          role="button"
          tabindex="-1"
        >
          <div
            class="mx_SpaceButton_selectionWrapper"
          >
            <div
              class="mx_SpaceButton_avatarWrapper"
            >
              <span
                class="mx_BaseAvatar"
                role="presentation"
              >
                <span
                  aria-hidden="true"
                  class="mx_BaseAvatar_initial"
                  style="font-size: 20.8px; width: 32px; line-height: 32px;"
                >
                  R
                </span>
                <img
                  alt=""
                  aria-hidden="true"
                  class="mx_BaseAvatar_image"
                  data-testid="avatar-img"
                  loading="lazy"
                  src="data:image/png;base64,00"
                  style="width: 32px; height: 32px;"
                />
              </span>
            </div>
            <div
              aria-expanded="false"
              aria-haspopup="true"
              aria-label="Space options"
              class="mx_AccessibleButton mx_SpaceButton_menuButton"
              role="button"
              tabindex="0"
            />
          </div>
        </div>
      </li>
      <li
        aria-selected="false"
        class="mx_SpaceItem mx_SpaceItem_new collapsed"
        role="treeitem"
      >
        <div
          aria-label="Create a space"
          class="mx_AccessibleButton mx_SpaceButton mx_SpaceButton_new mx_SpaceButton_narrow"
          data-testid="create-space-button"
          role="button"
          tabindex="-1"
        >
          <div
            class="mx_SpaceButton_selectionWrapper"
          >
            <div
              class="mx_SpaceButton_avatarWrapper"
            >
              <div
                class="mx_SpaceButton_avatarPlaceholder"
              >
                <div
                  class="mx_SpaceButton_icon"
                />
              </div>
            </div>
          </div>
        </div>
      </li>
    </ul>
    <div
      aria-expanded="false"
      aria-label="Quick settings"
      class="mx_AccessibleButton mx_QuickSettingsButton"
      role="button"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;
